src/uct/api/uct.h

```diff
@@ -1366,10 +1366,15 @@ typedef struct uct_rkey_bundle {
  *    without the need to wait for completion.
  *  - If the number of operations is smaller than the initial value of the counter,
  *    the callback will not be called at all, so it may be left undefined.
+ *  - status field is required to track first time occurred error and report it
+ *    by user callback on last completion.
  */
 struct uct_completion {
     uct_completion_callback_t func;    /**< User callback function */
     int                       count;   /**< Completion counter */
+    ucs_status_t              status;  /**< Completion status, this field has to
+                                            be initialized with UCS_OK before
+                                            first operation is started. */
```

Comment:
Why initialization of the status required ? thanks

Reply:
actually, all fields have to be initialized

Reply:
@yosefe is there is no check for the first invocation ?

Reply:
Hmm, anyways you are going to overwrite it on an event ? And when the event is generated you do know the status current event, so it is not not obvious why it has to be set to "OK" before the event.

Reply:
the previous value will be used, like this, to contain the first error:
```
uct_invoke_completion(comp, status)
{
  if (comp->status == UCS_OK) {
     comp->status = status;
  }
  comp->count--;
  if (comp->count == 0) {
   comp->func(comp)
  }
}
```

Reply:
got it

Reply:
Can the func be NULL ?

Reply:
> Hmm, anyways you are going to overwrite it on an event ? And when the event is generated you do know the status current event, so it is not not obvious why it has to be set to "OK" before the event.

Actually there is a compile time optimization for "likely" fast path `uct_invoke_completion(comp, UCS_OK);`, in this case newly added branch will be compiled out from inline function.

Reply:
Based on the current code the function cannot be NULL. Let's fix this part of documentation as well.

Reply:
if user dont want to invoke callback it can be NULL but this is his responsibility don't allow count to reach 0, I think it's already documented here https://github.com/openucx/ucx/blob/bd4374aad190ddcbc0d5136decc8574c72408f02/src/uct/api/uct.h#L1362
"When the operation completes, "count" field is decremented by 1, and whenever it reaches 0 - the callback is called."
For example, user can initialize count to 3, post 2 operation and wait count == 1

Reply:
Valid point

Reply:
extra space at start of line 1360